eng/pipelines/templates/steps/build.yml

```diff
@@ -0,0 +1,200 @@
+  - ${{ if eq(parameters.ServiceDirectory, 'auto') }}:
+    - task: DownloadPipelineArtifact@2
+      inputs:
+        artifact: BuildPackagesArtifact
+        path: $(Build.ArtifactStagingDirectory)/PackageInfo
+  - ${{ else }}:
+    - template: /eng/common/pipelines/templates/steps/save-package-properties.yml
+      parameters:
+        ServiceDirectory: ${{ parameters.ServiceDirectory }}
+
+  - template: /eng/common/pipelines/templates/steps/daily-dev-build-variable.yml
+
+  - template: /eng/pipelines/templates/steps/set-artifact-packages.yml
+    parameters:
+      PackageInfo: '$(Build.ArtifactStagingDirectory)/PackageInfo'
+      Artifacts: ${{ parameters.Artifacts }}
+
+  - pwsh: |
+      # Reset SkipDevBuildNumber as empty
+      $skipDevBuildNumber = ""
+      echo "##vso[task.setvariable variable=SkipDevBuildNumber]$skipDevBuildNumber"
+      Write-Host "SkipDevBuildNumber: [$(SkipDevBuildNumber)]"
+      $skipDevBuildNumber = "false"
+      # For .NET builds the only case we want to not have dev build numbers on our packages is when it is manually queued and
+      # the SetDevVersion property isn't set to true. All other cases should contain dev version numbers.
+      if ('$(Build.Reason)' -eq 'Manual' -and '$(SetDevVersion)' -ne 'true') {
+        $skipDevBuildNumber = "true"
+      }
+      $versioningProperties = "/p:OfficialBuildId=$(OfficialBuildId) /p:SkipDevBuildNumber=$skipDevBuildNumber";
+      echo "##vso[task.setvariable variable=VersioningProperties]$versioningProperties"
+    displayName: "Setup .NET specific versioning properties"
+
+  - task: UsePythonVersion@0
+    displayName: 'Use Python 3.11'
+    inputs:
+      versionSpec: '3.11'
+
+  - template: /eng/pipelines/templates/steps/dotnet-diagnostics.yml
+    parameters:
+      LogFilePath: $(Build.ArtifactStagingDirectory)/pack.binlog
+
+  # only use the project list override file if the service directory is not auto
+  - ${{ if eq(parameters.ServiceDirectory, 'auto') }}:
+    - pwsh: |
+        if (!"$(ProjectNames)") {
+          Write-Host "No packages to build."
+          exit 0
+        }
+        dotnet pack eng/service.proj -warnaserror `
+          /p:ValidateRunApiCompat=true `
+          /p:SDKType=${{ parameters.SDKType }} `
+          /p:IncludeTests=false `
+          /p:PublicSign=false $(VersioningProperties) `
+          /p:Configuration=$(BuildConfiguration) `
+          /p:CommitSHA=$(Build.SourceVersion) `
+          /p:ArtifactsPackagesDir=$(Build.ArtifactStagingDirectory) `
+          /p:ServiceDirectory=* `
+          /p:ProjectListOverrideFile=$(ProjectListOverrideFile) `
+          /p:IncludePerf=false `
+          /p:IncludeStress=false `
+          /p:IncludeIntegrationTests=false `
+          /p:EnableOverrideExclusions=true `
+          $(DiagnosticArguments)
+      displayName: "Build and Package for PR"
+  - ${{ else }}:
+    - pwsh: |
+        $variableSet = $env:ProjectListOverrideFile -ne "`$`(ProjectListOverrideFile`)"
```

Comment:
`$env:projectListOverrideFile` should be null/emtpty if it isn't set. So you shouldn't need the fancy check you are using here.

Reply:
Unfortunately, I do need to either A) clean it up or b) never set it.

The reason being, it being around means it gets picked up by anything later in `build` that uses the traversal project, because of how `msbuild` honors environment variables.

Reply:
I think I got it now with changes to `set-artifact-packages.ps1`.